File: apps/frontend/lib/types/photo.ts
/**
 * Resume Photo Configuration — frontend contract (Photo System).
 *
 * Mirrors the backend `app/profile/photo.py` 1:1. A resume's photo is described
 * by a structured {@link PhotoConfig} (presentation + provenance), never a
 * boolean and never a copy of the image. The config lives in
 * `resumeData.personalInfo.photo`; the resolved URL lives in
 * `personalInfo.avatarUrl`.
 *
 * Presentation vs. provenance are separated on purpose:
 * - Presentation (shape/size/position/crop/offset/zoom/frame): pure render hints
 *   templates map to their own layout via {@link TemplatePhotoCapability}.
 * - Provenance (`ref` + `snapshot`): `canonical` tracks the live profile photo;
 *   `snapshot` is pinned to the master captured at generation time (immune to a
 *   later profile-photo change).
 */

export type PhotoShape = 'circle' | 'rounded' | 'square' | 'custom';
export type PhotoSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'custom';
export type PhotoAlign = 'left' | 'center' | 'right';
export type PhotoPosition =
  | 'template-default'
  | 'header-left'
  | 'header-right'
  | 'header-center'
  | 'sidebar'
  | 'floating';
export type PhotoCrop = 'cover' | 'contain' | 'fill';
export type PhotoRef = 'canonical' | 'snapshot';

export interface PhotoSnapshot {
  url?: string | null;
  checksum?: string | null;
  width?: number | null;
  height?: number | null;
}

export interface PhotoConfig {
  show: boolean;
  ref: PhotoRef;
  snapshot: PhotoSnapshot;
  // Presentation
  shape: PhotoShape;
  radius: number; // px, when shape === 'custom'
  size: PhotoSize;
  customSize?: number | null; // px, when size === 'custom'
  align: PhotoAlign;
  position: PhotoPosition;
  crop: PhotoCrop;
  offsetX: number; // object-position % (0-100)
  offsetY: number;
  zoom: number; // >= 1 scales the master up inside the fixed frame
  border: boolean;
  borderWidth: number;
  borderColor: string;
  shadow: boolean;
  background?: string | null;
  opacity: number; // 0-1
  margin: number; // px around the frame
}

/** Size token → rendered edge length in px (must match backend SIZE_PX). */
export const PHOTO_SIZE_PX: Record<Exclude<PhotoSize, 'custom'>, number> = {
  xs: 48,
  sm: 64,
  md: 96,
  lg: 128,
  xl: 160,
};

/** Safe default: hidden, tracking the canonical profile photo, circular, medium. */
export const DEFAULT_PHOTO_CONFIG: PhotoConfig = {
  show: false,
  ref: 'canonical',
  snapshot: {},
  shape: 'circle',
  radius: 12,
  size: 'md',
  customSize: null,
  align: 'left',
  position: 'template-default',
  crop: 'cover',
  offsetX: 50,
  offsetY: 50,
  zoom: 1,
  border: false,
  borderWidth: 2,
  borderColor: '#e5e7eb',
  shadow: false,
  background: null,
  opacity: 1,
  margin: 0,
};

const clamp = (v: number, lo: number, hi: number) => Math.max(lo, Math.min(hi, v));

/** Normalize a partial/legacy config into a complete, clamped PhotoConfig. */
export function normalizePhotoConfig(input?: Partial<PhotoConfig> | null): PhotoConfig {
  if (!input) return { ...DEFAULT_PHOTO_CONFIG };
  return {
    ...DEFAULT_PHOTO_CONFIG,
    ...input,
    snapshot: { ...DEFAULT_PHOTO_CONFIG.snapshot, ...(input.snapshot ?? {}) },
    offsetX: clamp(input.offsetX ?? 50, 0, 100),
    offsetY: clamp(input.offsetY ?? 50, 0, 100),
    zoom: clamp(input.zoom ?? 1, 1, 3),
    opacity: clamp(input.opacity ?? 1, 0, 1),
  };
}

/** Rendered edge length in px for a config's size token. */
export function resolvedSizePx(config: PhotoConfig): number {
  if (config.size === 'custom' && config.customSize) {
    return clamp(Math.round(config.customSize), 24, 512);
  }
  return PHOTO_SIZE_PX[config.size as Exclude<PhotoSize, 'custom'>] ?? PHOTO_SIZE_PX.md;
}

/**
 * Resolve the URL a resume should render (provenance-aware). Single source of
 * truth mirroring the backend `resolve_photo_url`.
 */
export function resolvePhotoUrl(
  config: PhotoConfig | null | undefined,
  profileAvatarUrl: string | null | undefined
): string | null {
  if (!config || !config.show) return null;
  if (config.ref === 'snapshot') return config.snapshot?.url || profileAvatarUrl || null;
  return profileAvatarUrl || null;
}

/** CSS border-radius for a shape. */
export function shapeRadius(config: PhotoConfig): string {
  switch (config.shape) {
    case 'circle':
      return '9999px';
    case 'square':
      return '0px';
    case 'rounded':
      return '12px';
    case 'custom':
      return `${Math.max(0, config.radius)}px`;
    default:
      return '9999px';
  }
}

File: apps/backend/app/profile/photo.py
"""Resume Photo Configuration — the render+provenance contract (Photo System).

A resume's photo is described by a structured :class:`PhotoConfig`, **not** a
boolean and **not** a copy of the image. It lives in the resume's
``processed_data.personalInfo.photo`` (co-located with the header identity it
decorates) and captures every render decision the templates honour: visibility,
shape, size, alignment, position slot, crop mode, in-frame reposition/zoom,
border/shadow/background, opacity, and margin.

Two orthogonal concerns are separated on purpose:

1. **Presentation** (shape/size/position/…): pure render hints. Templates map
   the *position slot* + *size token* to their own layout; they never hardcode a
   pixel layout, and unknown slots fall back to the template default.
2. **Provenance** (``ref`` + ``snapshot``): where the pixels come from.
   - ``ref="canonical"`` → the resume tracks the user's *live* profile photo.
     Replacing the profile photo later updates this resume too (opt-in freshness).
   - ``ref="snapshot"`` → the resume is pinned to the exact master captured at
     generation time (``snapshot.url``/checksum/dims). A later profile-photo
     change never mutates this already-generated resume.

This module is **pure** (no I/O), so it is shared by the Projection Engine,
sync, public projection, and tests, and it mirrors the frontend
``lib/types/photo.ts`` type 1:1.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, field_validator

__all__ = ["PhotoConfig", "PhotoSnapshot", "resolve_photo_url", "DEFAULT_PHOTO_CONFIG"]

PhotoShape = Literal["circle", "rounded", "square", "custom"]
PhotoSize = Literal["xs", "sm", "md", "lg", "xl", "custom"]
PhotoAlign = Literal["left", "center", "right"]
PhotoPosition = Literal[
    "template-default",
    "header-left",
    "header-right",
    "header-center",
    "sidebar",
    "floating",
]
PhotoCrop = Literal["cover", "contain", "fill"]
PhotoRef = Literal["canonical", "snapshot"]

# Size token → rendered edge length in px (the render layer maps these; kept
# here so preview and PDF agree and future surfaces reuse one scale).
SIZE_PX: dict[str, int] = {"xs": 48, "sm": 64, "md": 96, "lg": 128, "xl": 160}


class PhotoSnapshot(BaseModel):
    """Frozen master reference for a pinned (``ref="snapshot"``) resume photo."""

    url: str | None = None
    checksum: str | None = None
    width: int | None = None
    height: int | None = None


class PhotoConfig(BaseModel):
    """Per-resume photo configuration (presentation + provenance)."""

    show: bool = False
    ref: PhotoRef = "canonical"
    snapshot: PhotoSnapshot = Field(default_factory=PhotoSnapshot)

    # Presentation ---------------------------------------------------------
    shape: PhotoShape = "circle"
    radius: int = 12  # px, only used when shape == "custom"
    size: PhotoSize = "md"
    customSize: int | None = None  # px, only used when size == "custom"
    align: PhotoAlign = "left"
    position: PhotoPosition = "template-default"
    crop: PhotoCrop = "cover"
    # In-frame reposition (object-position %) + zoom (>=1 scales the master up
    # inside the fixed frame). Together these give crop/recenter/resize with zero
    # mutation of the master.
    offsetX: float = 50.0
    offsetY: float = 50.0
    zoom: float = 1.0
    # Framing.
    border: bool = False
    borderWidth: int = 2
    borderColor: str = "#e5e7eb"
    shadow: bool = False
    background: str | None = None
    opacity: float = 1.0
    margin: int = 0  # px around the frame

    @field_validator("opacity")
    @classmethod
    def _clamp_opacity(cls, v: float) -> float:
        return max(0.0, min(1.0, float(v)))

    @field_validator("offsetX", "offsetY")
    @classmethod
    def _clamp_offset(cls, v: float) -> float:
        return max(0.0, min(100.0, float(v)))

    @field_validator("zoom")
    @classmethod
    def _clamp_zoom(cls, v: float) -> float:
        # Guard against absurd zoom (perf + layout). 1.0 = fit, 3.0 = tight crop.
        return max(1.0, min(3.0, float(v)))

    def resolved_size_px(self) -> int:
        """Rendered edge length in px for the current size token."""
        if self.size == "custom" and self.customSize:
            return max(24, min(512, int(self.customSize)))
        return SIZE_PX.get(self.size, SIZE_PX["md"])


# Safe default: photo hidden, tracking the canonical profile photo, circular,
# medium, header-left. Turning a photo on is an explicit, per-resume opt-in.
DEFAULT_PHOTO_CONFIG = PhotoConfig()


def resolve_photo_url(config: PhotoConfig | None, profile_avatar_url: str | None) -> str | None:
    """Resolve the URL a resume should render for its photo (provenance-aware).

    - Hidden / no config → ``None`` (template renders its no-photo fallback).
    - ``ref="snapshot"`` → the pinned ``snapshot.url`` (frozen; immune to a later
      profile-photo change).
    - ``ref="canonical"`` → the *live* ``profile_avatar_url`` (tracks the profile).

    This single function is the authority reused by projection, read-time
    re-resolution, and public projection so the rule is enforced in one place.
    """
    if config is None or not config.show:
        return None
    if config.ref == "snapshot":
        return config.snapshot.url or profile_avatar_url
    return profile_avatar_url
